### PR TITLE
feat(hostRules): support https options and platform in host rules from env

### PR DESCRIPTION
**Pull request by lyonlai**
## Changes

made `hostRulesFromEnv` able to:

* pick up https options from environment variables.
* pick up credentials for platform in additional to datasource

## Context
As per changes for [discussion of moving setGlobalHostRules before platform initialisation](https://github.com/renovatebot/renovate/discussions/24660). Renovate needs to be able to pickup [https options](https://github.com/renovatebot/renovate/pull/24155) from environment variable and apply before platform initialisation.

## Documentation (please check one with an [x])

- [X] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/types/host-rules.ts
Comment (rarkins) [inline lib/types/host-rules.ts:1]: Avoid using any

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:269]: We need to test thoroughly that this doesn't cause unintended logs or errors if variables like GITHUB_TOKEN are in use

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/types/host-rules.ts
Comment (lyonlai) [reply lib/types/host-rules.ts:1]: what would you suggest instead of `any` here? would `unknown` works?

Comment (lyonlai) [reply lib/types/host-rules.ts:1]: the reason behind this line change here is tsc will complain on the `rule[suffix]` references in [here](https://github.com/lyonlai/renovate/blob/main/lib/workers/global/config/parse/host-rules-from-env.ts#L31)

```
lib/workers/global/config/parse/host-rules-from-env.ts:61:9 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'HostRule'.
  No index signature with a parameter of type 'string' was found on type 'HostRule'.

61         rule[suffix] = env[envName];
```

would be appreciated for any better way forward.

File: docs/usage/self-hosted-configuration.md
Comment (lyonlai) [reply docs/usage/self-hosted-configuration.md:269]: sure. I have made a change and make a unit test for it.

File: lib/types/host-rules.ts
Comment (lyonlai) [reply lib/types/host-rules.ts:1]:
```diff
diff --git a/lib/modules/datasource/packagist/index.spec.ts b/lib/modules/datasource/packagist/index.spec.ts
index e445de179..a4a02f7a8 100644
--- a/lib/modules/datasource/packagist/index.spec.ts
+++ b/lib/modules/datasource/packagist/index.spec.ts
@@ -2,7 +2,8 @@ import { mockDeep } from 'jest-mock-extended';
 import { getPkgReleases } from '..';
 import { Fixtures } from '../../../../test/fixtures';
 import * as httpMock from '../../../../test/http-mock';
-import type { HostRule } from '../../../types';
+import type { HostRuleSearchResult } from '../../../types';
+import type { HostRuleSearch } from '../../../util/host-rules';
 import * as _hostRules from '../../../util/host-rules';
 import * as composerVersioning from '../../versioning/composer';
 import { id as versioning } from '../../versioning/loose';
@@ -25,7 +26,7 @@ describe('modules/datasource/packagist/index', () => {
     let config: any;

     beforeEach(() => {
-      hostRules.find = jest.fn((input: HostRule) => input);
+      hostRules.find = jest.fn((input: HostRuleSearch) => input as HostRuleSearchResult);
       hostRules.hosts = jest.fn(() => []);
       config = {
         versioning: composerVersioning.id,
diff --git a/lib/types/host-rules.ts b/lib/types/host-rules.ts
index 6206b9c81..bb9d689aa 100644
--- a/lib/types/host-rules.ts
+++ b/lib/types/host-rules.ts
@@ -1,4 +1,4 @@
-export interface HostRuleSearchResult extends Record<string, any> {
+export interface HostRuleSearchResult extends Record<string, unknown> {
   authType?: string;
   token?: string;
   username?: string;
```

I've tried locally with `unknown`. There were some tiny issue with test mocking in datasource. Here is the diff. Let me know if it is better. @rarkins

Comment (rarkins) [reply lib/types/host-rules.ts:1]: @viceice @zharinov would some "key of" TS trickery work here?

Comment (lyonlai) [reply lib/types/host-rules.ts:1]: I have tried to use `keyof` tricks on the object index
both `rule[suffix as keyof typeof rule] = env[envName]` or `rule[suffix as keyof HostRule] = env[envName]` gives me `Type 'string | undefined' is not assignable to type undefined`

File: docs/usage/self-hosted-configuration.md
Comment (lyonlai) [reply docs/usage/self-hosted-configuration.md:269]: do you think the change I made is sufficient to cover the situation (e.g. GITHUB_TOKEN) you mentioned?  @rarkins . let me know if there other things I could help

File: lib/types/host-rules.ts
Comment (viceice) [reply lib/types/host-rules.ts:1]: we can't simply copy that values, as some of them need to be numbers. so you need a switch case. that will also solve the typescript issue.

File: lib/workers/global/config/parse/host-rules-from-env.ts
Comment (viceice) [inline lib/workers/global/config/parse/host-rules-from-env.ts:38]: bad idea. this gives transitive circular reference which can cause runtime issues. why do you need the host types of all platforms?

File: lib/workers/global/config/parse/host-rules-from-env.ts
Comment (viceice) [inline lib/workers/global/config/parse/host-rules-from-env.ts:64]: we should only copy if `env[envName]` isn't undefined.

File: lib/types/host-rules.ts
Comment (lyonlai) [reply lib/types/host-rules.ts]: that's a good one. thank you. I've wrapped it up in a function

File: lib/workers/global/config/parse/host-rules-from-env.ts
Comment (lyonlai) [reply lib/workers/global/config/parse/host-rules-from-env.ts]: I was just following the patterns. I don't need other host types, I've made it just for github for now.

File: lib/workers/global/config/parse/host-rules-from-env.ts
Comment (lyonlai) [reply lib/workers/global/config/parse/host-rules-from-env.ts]: that's done in the `setHostRuleValue` function

File: docs/usage/self-hosted-configuration.md
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:269]: Do you mean this line? `(platforms.has(hostType) && splitEnv.length > 1)`

i.e. it needs to be `GITHUB_X_TOKEN` or longer?

Comment (lyonlai) [reply docs/usage/self-hosted-configuration.md:269]: I was intended to just to avoid processing environment variable like `{{PLATFORM}}_TOKEN`. would you like to just skip those environment variable from a list ?

Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:269]: I think it's ok to skip all of them, although best to document that fact

Comment (lyonlai) [reply docs/usage/self-hosted-configuration.md:269]: cool. I've updated the doc.

**Comment by lyonlai** [comment]
@rarkins and @viceice would you mind take an other look and let me know if there's anything else that I need to do to get this in?

**Review by rarkins — DISMISSED** [review]

File: docs/usage/self-hosted-configuration.md
Comment (astellingwerf) [inline docs/usage/self-hosted-configuration.md:280]: The standard is to have one sentence per line in Markdown

Comment (lyonlai) [reply docs/usage/self-hosted-configuration.md]: sure. I've put it in a new line.

**Review by rarkins — DISMISSED** [review]

**Comment by lyonlai** [comment]
@rarkins 👋  how are we looking for the merge? are we still waiting for anything?

**Comment by rarkins** [comment]
Awaiting second review

**Comment by lyonlai** [comment]
@viceice would you take a second review on this please?

**Comment by lyonlai** [comment]
@rarkins there seems to be no progress on the second review for the past week. Any other member that could perform a second review to get this going?

**Comment by rarkins** [comment]
@viceice can you merge if this looks ok to you?

File: lib/workers/global/config/parse/host-rules-from-env.ts
Comment (viceice) [inline lib/workers/global/config/parse/host-rules-from-env.ts:48]:
```suggestion
        rule[key] = value;
```

File: docs/usage/self-hosted-configuration.md
Comment (viceice) [inline docs/usage/self-hosted-configuration.md:269]:
```suggestion
- Datasource name (e.g. `NPM`, `PYPI`) or Platform name (only  `GITHUB`)
```
you only support GitHub.

File: lib/workers/global/config/parse/host-rules-from-env.ts
Comment (viceice) [inline lib/workers/global/config/parse/host-rules-from-env.ts:77]:
```suggestion
      if (isHttpsAuthField(suffix)) {
        suffix = restoreHttpsAuthField(suffix);
```

remove unnecessary checks

**Review by rarkins — APPROVED** [review]

**Pull request merged**
